tests/scripts/all.sh

```diff
@@ -384,7 +384,9 @@ $text"
     }
 else
     record_status () {
-        "$@"
+        if "$@"; then :; else
```

Comment:
I don't understand this change. It doesn't help to make `record_status ! $command` work. And any change to `record_status` for this purpose needs to also be applied to the keep-going implementation.

Here are several possible solutions:

1. Add `!` detection code to both `record_status` implementations. Replace `"$@"` by

        if [ "$1" = "!" ]; then shift; ! "$@"; else "$@"; fi

2. Instead of `record_status ! …`, use `record_status not …` with

        not () { ! "$@"; }

  (`!` is not a valid function name in many sh implementations.)

3. Instead of `record_status ! …`, use `record_status_not …`, with `record_status_not` implemented for both keep-going and non-keep-going mode.

4. Put the negation in a function:

        check_absence () {
            ! grep "$@"
        }
        if_build_succeeded check_absence __aeabi_uldiv library/*.o

I don't like the duplication in (1) and (3). I have no strong preference between (2) and (4).

Reply:
@gilles-peskine-arm This was a misguided attempt as solving the problem, that I should have tested correctly before pushing (I tested it after pushing, but then had no time to fix it before I left the office). I was under the impression, from previous testing, that the "keep going" implementation was working, so I figured `!` only worked in the context of an `if` and decided to aligned to that. Obviously I was very mistaken.

I'll switch to the `not` solution, which I slightly prefer to the `check_absence` one as it is more generic.

Reply:
On second thoughts `not` perpetuates the defect that `grep` could possibly fail for other reasons (very unlikely ones such as a disk error). A more reliable test would be
```
check_absence () {
     ( set +e; grep "$@" ); [ $? -eq 1 ]
 }
```
or
```
check_absence () {
     [ -z "$(grep -l "$@")" ]
 }
```
But I'm fine with `not` too, since I was happy enough with it earlier to commit it and in any case it would be a preexisting issue.